src/probnum/_config.py

```diff
@@ -0,0 +1,83 @@
+import contextlib
+from typing import Any
+
+
+class Configuration:
+    """
+    Configuration over which some mechanics of probnum can be controlled dynamically.
+
+    ``probnum`` provides some configurations together with default values. These
+    are listed in the tables below.
+    Additionally, the user can register own configuration entries via the method
+    :meth:`register`. Configuration entries can only be registered once and can only
+    be used (accessed or overwritten) once they have been registered.
+
+    - ``probnum.randvars``
+
+    +----------------------------------+-----------------------------------------------+
+    | Config entry                     | Description                                   |
+    +==================================+===============================================+
+    | ``covariance_inversion_damping`` | A (typically small) value that is per         |
+    |                                  | default added to the diagonal of covariance   |
+    |                                  | matrices in order to make inversion           |
+    |                                  | numerically stable.                           |
+    +----------------------------------+-----------------------------------------------+
+    | ``...``                          | ...                                           |
+    +----------------------------------+-----------------------------------------------+
+
+    - ``probnum.diffeq``
+
+    - ``...``
+
+
+    Examples
+    ========
+
+    >>> import probnum
+    >>> probnum.config.covariance_inversion_damping
+    1e-12
+    >>> with probnum.config(
+    ...     covariance_inversion_damping=1e-2,
+    ... ):
+    ...     probnum.config.covariance_inversion_damping
+    0.01
+    """
+
+    @contextlib.contextmanager
+    def __call__(self, **kwargs) -> None:
+        old_entries = dict()
+
+        for key, value in kwargs.items():
+            if not hasattr(self, key):
+                raise KeyError(
+                    f"Configuration entry {key} does not exist yet."
+                    "Configuration entries must be `register`ed before they can be "
+                    "accessed."
+                )
+
+            old_entries[key] = getattr(self, key)
+
+            setattr(self, key, value)
+
+        try:
+            yield
+        finally:
+            self.__dict__.update(old_entries)
+
+    def __setattr__(self, key: str, value: Any) -> None:
+        if not hasattr(self, key):
+            raise KeyError(
+                f"Configuration entry {key} does not exist yet."
+                "Configuration entries must be `register`ed before they can be "
+                "accessed."
+            )
+
+        self.__dict__[key] = value
```

Comment:
just to clarify my understanding: the difference between this and `settatr(self, key, value)` is that settatr does not throw errors? If not, what is the difference? (this is no real review-comment, but I'd like to learn this :))

Reply:
To the best of my knowledge, this is pretty much all the difference, yes. 

**To make this a review-comment, however:**
We just added this additional check in order to forbid setting random configs from outside, without registering that. 
Although this is as well debatable, imo. So far I am not sure what benefit the need for `register`ing config options brings. @marvinpfoertner has to answer this, since I did not implement this.

My opinion on this: Either forbid the user to add their own options or get rid of the `register`.

Reply:
The idea behind the `register` is to force `probnum` developers to register the "semantics" of a configuration option "statically", by which I mean as module-level code which gets executed while importing `probnum`.
This will prevent random scattered writes to the `pn.config` and provide central places in the code, where we get an overview of all registered probnum configuration options

Also, having a separate `register` call allows us to document configuration options automatically in the future.

A bit of context: The idea behind first registering (with a default value), then writing/setting/changing a configuration option comes from experience I had with a similar system, where people started to write new config keys in the middle of the library and in the end, nobody knew that certain config options already existed. It was a mess.

Reply:
>  people started to write new config keys in the middle of the library

I see what you mean. But cannot then people just add random `.register()` calls in the middle of nowhere, instead? Since register does not add any functionality, other than checking for duplicate `register`s

Reply:
Yes that's true, but I think the threshold for actually doing that is much higher than just writing something like `config.option = value`.